lib/ecto/adapters/mssql/connection.ex

```diff
@@ -0,0 +1,1644 @@
+      end)
+    end
+
+    defp expr({:datetime_add, _, [datetime, count, interval]}, sources, query) do
+      [
+        "DATEADD(",
+        interval,
+        ", ",
+        interval_count(count, sources, query),
+        ", CAST(",
+        expr(datetime, sources, query),
+        " AS datetime2(6)))"
+      ]
+    end
+
+    defp expr({:date_add, _, [date, count, interval]}, sources, query) do
+      "CAST(DATEADD(" <>
+        interval <>
+        ", " <>
+        interval_count(count, sources, query) <>
+        ", CAST(" <>
+        expr(
+          date,
+          sources,
+          query
+        ) <> " AS datetime2(6))" <> ") AS date)"
+    end
+
+    defp expr({:count, _, []}, _sources, _query), do: "count(*)"
+
+    defp expr({fun, _, args}, sources, query) when is_atom(fun) and is_list(args) do
+      {modifier, args} =
+        case args do
+          [rest, :distinct] -> {"DISTINCT ", [rest]}
+          _ -> {"", args}
+        end
+
+      case handle_call(fun, length(args)) do
+        {:binary_op, op} ->
+          [left, right] = args
+          [op_to_binary(left, sources, query), op | op_to_binary(right, sources, query)]
+
+        {:fun, fun} ->
+          [fun, ?(, modifier, intersperse_map(args, ", ", &expr(&1, sources, query)), ?)]
+      end
+    end
+
+    defp expr(list, sources, query) when is_list(list) do
+      Enum.map_join(list, ", ", &expr(&1, sources, query))
+    end
+
+    defp expr({string, :varchar}, _sources, _query)
+         when is_binary(string) do
+      "'#{escape_string(string)}'"
+    end
+
+    defp expr(string, _sources, _query) when is_binary(string) do
+      if String.contains?(string, @unsafe_query_strings) do
+        len = String.length(string)
+
+        hex =
+          string
+          |> :unicode.characters_to_binary(:utf8, {:utf16, :little})
+          |> Base.encode16(case: :lower)
+
+        "CONVERT(nvarchar(#{len}), 0x#{hex})"
+      else
+        "N'#{escape_string(string)}'"
+      end
+    end
+
+    defp expr(%Decimal{} = decimal, _sources, _query) do
+      Decimal.to_string(decimal, :normal)
+    end
+
+    defp expr(%Tagged{value: binary, type: :binary}, _sources, _query) when is_binary(binary) do
+      hex = Base.encode16(binary, case: :lower)
+      "0x#{hex}"
+    end
+
+    defp expr(%Tagged{value: binary, type: :uuid}, _sources, _query) when is_binary(binary) do
+      case binary do
+        <<_::64, ?-, _::32, ?-, _::32, ?-, _::32, ?-, _::96>> ->
+          {:ok, value} = Tds.Types.UUID.dump(binary)
+          value
+
+        any ->
+          any
+      end
+    end
+
+    defp expr(%Tagged{value: other, type: type}, sources, query)
+         when type in [:varchar, :nvarchar] do
+      "CAST(#{expr(other, sources, query)} AS #{column_type(type, [])}(max))"
+    end
+
+    defp expr(%Tagged{value: other, type: type}, sources, query) do
+      "CAST(#{expr(other, sources, query)} AS #{ecto_cast_to_db(type, [])})"
```

Comment:
`ecto_cast_to_db` and `column_type` are exactly the same function. We probably should keep only one of them.